unixfs/src/dir/builder/iter.rs

```diff
@@ -0,0 +1,416 @@
+                        let version_len = 1;
+                        let codec_len = sizeof_varint(u64::from(self.0.codec()));
+                        let hash_len = self.0.hash().as_bytes().len();
+                        version_len + codec_len + hash_len
+                    }
+                }
+            }
+
+            fn write_message<W: WriterBackend>(
+                &self,
+                w: &mut Writer<W>,
+            ) -> quick_protobuf::Result<()> {
+                use cid::Version::*;
+
+                match self.0.version() {
+                    V0 => {
+                        for b in self.0.hash().as_bytes() {
+                            w.write_u8(*b)?;
+                        }
+                        Ok(())
+                    }
+                    V1 => {
+                        // it is possible that Cidv1 should not be linked to from a unixfs
+                        // directory; at least go-ipfs 0.5 `ipfs files` denies making a cbor link
+                        w.write_u8(1)?;
+                        w.write_varint(u64::from(self.0.codec()))?;
+                        for b in self.0.hash().as_bytes() {
+                            w.write_u8(*b)?;
+                        }
+                        Ok(())
+                    }
+                }
+            }
+        }
+
+        impl<'a> MessageWrite for BTreeMappedDir<'a> {
+            fn get_size(&self) -> usize {
+                use quick_protobuf::sizeofs::*;
+
+                let links = self
+                    .links
+                    .iter()
+                    .map(|(k, v)| EntryAsPBLink(k, v))
+                    .map(|link| 1 + sizeof_len(link.get_size()))
+                    .sum::<usize>();
+
+                links + 1 + sizeof_len(self.data.get_size())
+            }
+            fn write_message<W: WriterBackend>(
+                &self,
+                w: &mut Writer<W>,
+            ) -> quick_protobuf::Result<()> {
+                for l in self.links.iter().map(|(k, v)| EntryAsPBLink(k, v)) {
+                    w.write_with_tag(18, |w| w.write_message(&l))?;
+                }
+                w.write_with_tag(10, |w| w.write_message(&self.data))
+            }
+        }
+
+        let btreed = BTreeMappedDir {
+            links,
+            data: UnixFs {
+                Type: UnixFsType::Directory,
+                Data: None,
+                ..Default::default()
+            },
+        };
+
+        let size = btreed.get_size();
+
+        // FIXME: we shouldn't be creating too large structures (bitswap block size limit!)
+        // FIXME: changing this to autosharding is going to take some thinking
+
+        let cap = buffer.capacity();
+
+        if let Some(additional) = size.checked_sub(cap) {
+            buffer.reserve(additional);
+        }
+
+        if let Some(needed_zeroes) = size.checked_sub(buffer.len()) {
+            buffer.extend(std::iter::repeat(0).take(needed_zeroes));
```

Comment:
```suggestion
            buffer.extend_from_slice(&[0; needed_zeroes]);
```

Reply:
I was thinking about this.

The block limit will be ~0.5MB or something like that. I don't think doing a 500kB stack allocation works at all times, or would this get optimized to something else? On linux that'd surely work. Here doing a `set_len` would be "safe" as well as quick-protobuf will write the size bytes (unless my CidMessageWrite fails).

Reply:
Leaving this out now. I bet that `set_len` will be needed if someone wants to use this on a smaller device.

Reply:
though ... doing this with a smaller, for example [0; 64] might be good. I should write some benchmarks.